### ISSUE TITLE
"No route to host" due to endpoint removal

### ISSUE THREAD

**Issue by barkbay** [openshift/openshift-sdn#281]
Hi all,

From time to time we lose some services. What I mean is that suddenly connections to a service fail with a "no route to host" error.
Interestingly, when it happens, we have these messages in the log files on the nodes :

```
12:01:16 node3-production-mt.digital.example.com origin-node[129471]: W0404 12:01:16.267651  129471 registry.go:638] Service 'srv1' in namespace 'ns1' has an Endpoint pointing to pod 10.1.2.8 in namespace 'ns2' 
12:01:16 node3-production-mt.digital.example.com origin-node[129471]: I0404 12:01:16.267945  129471 proxier.go:487] Removing endpoints for "ns1/srv1:web"
```

The point is that the pods with the address 10.1.2.8 in ns2 is an old build, it is not "active" anymore. I don't understand why it matters to Openshift or Kubernetes that a "dead" pod had this IP before.

It seems to happen since we have upgraded from Openshift v1.1.0.1 to Openshift v1.1.4/kubernetes v1.2.0-origin-41-g91d3e75 using these RPMs : https://copr-be.cloud.fedoraproject.org/results/maxamillion/origin-next/epel-7-x86_64/00168477-origin/
I don't know if it is relevant but we use the "multitenant" sdn plugin.

To be honest I would be grateful to anyone how can help me because we have this issue in production :(

Thanks

**Comment by barkbay** [comment]
Seems related to #285 
I will try backport and give it a try

**Comment by barkbay** [comment]
So, I made a patch for Openshit Origin 1.1.6 https://gist.github.com/barkbay/c86672f933d0bb6f9eaf38ec06cfe9ce

Here is a simple test : 

1. Start a build in a namespace and find the IP : 

  ```
  [openshift@lenaic-master-4a9cb ~]$ oc get pod/php3-1-build -n foo -o yaml|grep podIP 
  podIP: 10.1.1.3
  ```
2. Deploy some pods in an other namespace until the IP 10.1.1.3 is reused :
```
[root@lenaic-node-compute-2c3ad openshift]# journalctl -u origin-node |grep -e 'registry.go' -e 'proxier.go'|grep '10.1.1.3'
[.....]
11:25:54 lenaic-node-compute-2c3ad.example.com origin-node[28642]: I0410 11:25:54.856861   28642 registry.go:728] TrackPod IP/Status 10.1.1.3/Running in NS foo 
11:26:29 lenaic-node-compute-2c3ad.example.com origin-node[28642]: W0410 11:26:29.262954   28642 registry.go:724] IP '10.1.1.3' was marked as used by namespace 'foo' (pod '4ecb60c6-ff06-11e5-8086-5254003d6d33')... updating to namespace 'bar' (pod '0687bb7f-ff0f-11e5-8086-5254003d6d33') 
11:26:29 lenaic-node-compute-2c3ad.example.com origin-node[28642]: I0410 11:26:29.263004   28642 registry.go:728] TrackPod IP/Status 10.1.1.3/Running in NS bar 
11:26:29 lenaic-node-compute-2c3ad.example.com origin-node[28642]: I0410 11:26:29.589729   28642 proxier.go:484] Setting endpoints for "bar/bar2:8080-tcp" to [10.1.0.2:8080 10.1.1.3:8080 10.1.1.4:8080]
[.....]
```

The only caveat so far is that if I scale a pod to 0 then the pod is not "untracked"  : 
```
avril 10 11:42:07 lenaic-node-compute-2c3ad.example.com origin-node[28642]: I0410 11:42:07.541477   28642 registry.go:728] TrackPod IP/Status 10.1.1.6/Running in NS bar
avril 10 11:42:07 lenaic-node-compute-2c3ad.example.com origin-node[28642]: I0410 11:42:07.631712   28642 proxier.go:494] Removing endpoints for "bar/postgresqlbar2:postgresql"
```

Would be glad to have your feedback because this bug makes Openshift 1.1.6 pretty much unusable as soon as you have a few namespaces.

Thanks !

**Comment by danwinship** [comment]
last comment suggests #285 didn't completely fix this

**Comment by barkbay** [comment]
In the following function are you sure that pod.Status.PodIP has a value while the pod has been deleted ?
It would explain why the pod is not removed from registry.podsByIP

```
func (registry *Registry) unTrackPod(pod *kapi.Pod) {
	// Only delete if the pod ID is the one we are tracking (in case there is a failed or complete
	// pod lying around that gets deleted while there is a running pod with the same IP)
	if podInfo, ok := registry.podsByIP[pod.Status.PodIP]; ok && podInfo.UID == pod.UID {
		delete(registry.podsByIP, pod.Status.PodIP)
	}

	return
}
```

**Comment by knobunc** [comment]
@barkbay @danwinship: I am not sure that it won't have an IP address.  However, I'm not sure that it matters.  As soon as a new pod hits pending or running state with that same IP address it will take ownership of the IP.  It would be nice to clean out the tracking when a pod is deleted, but if we don't know the old IP, then we can't clean it.  And tracking that would necessitate a separate cache for pods IDs to IP... and that seems like overkill.

**Comment by barkbay** [comment]
Does your last comment means that the `unTrackPod(pod *kapi.Pod)` function could be removed from the registry ?
And also the `log.Warningf("IP '%s' was marked as used by....`  in `trackPod(pod *kapi.Pod) ?`
I'm not sure that this message is relevant if we know that the cache is not consistent.

**Comment by barkbay** [comment]
Up

**Comment by knobunc** [comment]
@barkbay I would rather remove them from the cache so that it doesn't unnecessarily bloat.  I am not opposed to removing the warning message, or at least making it informational.

**Comment by jcias** [comment]
@barkbay @knobunc @danwinship I've encountered similar behavior at Openshift 1.5.1. While this issue is opened for almost 2 years in this dead repository, anyone knows if it was fixed or addressed anyhow? Any reference or URL? I'd appreciate any help.

**Comment by jcias** [comment]
It looks like this was resolved actually somewhere between 1.5.1 and 1.7 (3.7) version as after we've upgraded to 3.7 we did not encounter such behavior any more. It appears this is a known bug which was described [here](https://access.redhat.com/solutions/3077711) and [here](https://bugzilla.redhat.com/show_bug.cgi?id=1462338), resolved by installing OpenShift enterprise version 3.5.5.31, most probably not resolved in upstream (community) 1.5 version.
